x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/action/UpdateIndexMigrationVersionAction.java

```diff
@@ -0,0 +1,191 @@
+            return indexName;
+        }
+    }
+
+    public static class TransportAction extends TransportMasterNodeAction<Request, UpdateIndexMigrationVersionResponse> {
+        private final MasterServiceTaskQueue<UpdateIndexMigrationVersionTask> updateIndexMigrationVersionTaskQueue;
+
+        @Inject
+        public TransportAction(
+            TransportService transportService,
+            ClusterService clusterService,
+            ThreadPool threadPool,
+            ActionFilters actionFilters,
+            IndexNameExpressionResolver indexNameExpressionResolver
+        ) {
+            super(
+                UpdateIndexMigrationVersionAction.NAME,
+                transportService,
+                clusterService,
+                threadPool,
+                actionFilters,
+                Request::new,
+                indexNameExpressionResolver,
+                UpdateIndexMigrationVersionResponse::new,
+                threadPool.executor(ThreadPool.Names.MANAGEMENT)
+            );
+            this.updateIndexMigrationVersionTaskQueue = clusterService.createTaskQueue(
+                "update-index-migration-version-task-queue",
+                Priority.HIGH,
+                UPDATE_INDEX_MIGRATION_VERSION_TASK_EXECUTOR
+            );
+        }
+
+        private static final SimpleBatchedExecutor<UpdateIndexMigrationVersionTask, Void> UPDATE_INDEX_MIGRATION_VERSION_TASK_EXECUTOR =
+            new SimpleBatchedExecutor<>() {
+                @Override
+                public Tuple<ClusterState, Void> executeTask(UpdateIndexMigrationVersionTask task, ClusterState clusterState) {
+                    return Tuple.tuple(task.execute(clusterState), null);
+                }
+
+                @Override
+                public void taskSucceeded(UpdateIndexMigrationVersionTask task, Void unused) {
+                    task.listener.onResponse(null);
+                }
+            };
+
+        static class UpdateIndexMigrationVersionTask implements ClusterStateTaskListener {
+            private final ActionListener<Void> listener;
+            private final int indexMigrationVersion;
+            private final String indexName;
+
+            UpdateIndexMigrationVersionTask(ActionListener<Void> listener, int indexMigrationVersion, String indexName) {
+                this.listener = listener;
+                this.indexMigrationVersion = indexMigrationVersion;
+                this.indexName = indexName;
+            }
+
+            ClusterState execute(ClusterState currentState) {
+                IndexMetadata.Builder indexMetadataBuilder = IndexMetadata.builder(currentState.metadata().getIndices().get(indexName));
+                indexMetadataBuilder.putCustom(
+                    MIGRATION_VERSION_CUSTOM_KEY,
+                    Map.of(MIGRATION_VERSION_CUSTOM_DATA_KEY, Integer.toString(indexMigrationVersion))
+                );
+                indexMetadataBuilder.version(indexMetadataBuilder.version() + 1);
```

Comment:
Following up on our conversation from yesterday @albertzaharovits .

The `DataStreamLifecycleService` uses the `Metadata.put` method with `incrementVersion = true`. Under the hood it uses [IndexMetadata.withIncrementedVersion](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/cluster/metadata/IndexMetadata.java#L974) that creates a copy of the metadata with version + 1. 

It feels backwards to use the builder and then make another copy of the `IndexMetadata` we just built so I opted to keep the `version + 1` logic. Let me know what you think.

Reply:
Thanks for considering it!